qurator/eynollah/processor.py

```diff
@@ -0,0 +1,59 @@
+from json import loads
+from pkg_resources import resource_string
+from tempfile import NamedTemporaryFile
+from os.path import join
+
+from ocrd import Processor
+from ocrd_modelfactory import page_from_file
+from ocrd_models.ocrd_page import to_xml
+from ocrd_utils import (
+    getLogger,
+    MIMETYPE_PAGE,
+    assert_file_grp_cardinality,
+    make_file_id
+)
+
+from .eynollah import Eynollah
+
+OCRD_TOOL = loads(resource_string(__name__, 'ocrd-tool.json').decode('utf8'))
+
+class EynollahProcessor(Processor):
+
+    def __init__(self, *args, **kwargs):
+        kwargs['ocrd_tool'] = OCRD_TOOL['tools']['ocrd-eynollah-segment']
+        kwargs['version'] = OCRD_TOOL['version']
+        super().__init__(*args, **kwargs)
+
+    def process(self):
+        LOG = getLogger('eynollah')
+        assert_file_grp_cardinality(self.input_file_grp, 1)
+        assert_file_grp_cardinality(self.output_file_grp, 1)
+        for n, input_file in enumerate(self.input_files):
+            page_id = input_file.pageId or input_file.ID
+            LOG.info("INPUT FILE %s (%d/%d) ", page_id, n + 1, len(self.input_files))
+            pcgts = page_from_file(self.workspace.download_file(input_file))
+            self.add_metadata(pcgts)
+            page = pcgts.get_Page()
+            page_image, _, _ = self.workspace.image_from_page(page, page_id, feature_filter='binarized')
+            file_id = make_file_id(input_file, self.output_file_grp)
+            with NamedTemporaryFile(buffering=0, suffix='.tif') as f:
+                page_image.save(f.name)
+                eynollah_kwargs = {
+                    'dir_models': self.resolve_resource(self.parameter['models']),
+                    'allow_enhancement': self.parameter['allow_enhancement'],
+                    'curved_line': self.parameter['curved_line'],
+                    'full_layout': self.parameter['full_layout'],
+                    'allow_scaling': self.parameter['allow_scaling'],
+                    'headers_off': self.parameter['headers_off'],
+                    'override_dpi': self.parameter['dpi'] if self.parameter['dpi'] > 0 else None,
+                    'logger': LOG,
+                    'pcgts': pcgts,
+                    'image_filename': f.name}
+                Eynollah(**eynollah_kwargs).run()
+                self.workspace.add_file(
+                    ID=file_id,
+                    file_grp=self.output_file_grp,
+                    pageId=page_id,
+                    mimetype=MIMETYPE_PAGE,
+                    local_filename=join(self.output_file_grp, file_id) + '.xml',
+                    content=to_xml(pcgts))
```

Comment:
This is quite un-ocrd-ish IMO. It will unnecessarily **rebase** the output PAGE `@imageFilename` onto the derived image requested here – which will be a newly generated PIL.Image object for whatever the preceding workflow produced for the page level (except for the binarization).

I think that the `Eynollah` class should be generalized to some in-memory API for image and resulting PAGE, so both use-cases, the standalone CLI and the OCR-D CLI, fit well.

Reply:
Note: creating a new base image in that way also looses all meta-data from the original image (which might be needed by follow-up processors).

Idea: As long as we have no good integration, we could at least use an existing image file in the filesystem instead of creating a new one. (This could even still be a derived image. With the drawback that then still all image meta-data would be lost, due to https://github.com/OCR-D/core/issues/343.)

So how about picking the last non-binarized `AlternativeImage/@filename` or `Page/@imageFilename` instead of `self.workspace.image_from_page().save(fname)`?

Reply:
Thanks for the suggestion. As all the image reading logic is in the `Eynollah.imread` method and it caches the images (in `self._imgs`) it should not be too difficult to pre-populate that cache with the right conversion from the PIL image from `image_from_page` - that is why we refactored it that way, I just wanted to get this shipped asap, hence the NamedTemporaryFile hack. Will implement this without file system call today.